### PR TITLE
Fixes #29213 - Support el8

### PR DESCRIPTION
**Pull request by wbclark**

### CODE REVIEW

File: spec/spec_helper_acceptance.rb [hunk 1]
```diff
@@ -22,14 +22,15 @@
...
+      if fact_on(host, 'os.family') == 'RedHat'
         # don't delete downloaded rpm for use with BEAKER_provision=no +
         # BEAKER_destroy=no
         on host, 'sed -i "s/keepcache=.*/keepcache=1/" /etc/yum.conf'
         # refresh check if cache needs refresh on next yum command
         on host, 'yum clean expire-cache'
+        major = fact_on(host, 'os.release.major')
         # We need SCL as well for dynflow
-        if fact_on(host, 'operatingsystem') == 'CentOS'
+        if fact_on(host, 'operatingsystem') == 'CentOS' && major == '8'
```
Comment (ehelms) [inline spec/spec_helper_acceptance.rb:33]: `== 7` is what we want here :)

Comment (ekohl) [reply spec/spec_helper_acceptance.rb:33]: I'll take a look at incorporating that in https://github.com/theforeman/puppet-foreman_proxy/pull/581

**Comment by ehelms** [comment]
For the installer itself running against a system, the only error I hit with default config is in this module:

```
  [ERROR 2020-04-20T19:43:51 verbose]  /Stage[main]/Foreman_proxy::Tftp::Netboot/File[/var/lib/tftpboot/grub2/shim.efi]: Could not evaluate: Could not retrieve information from environment production source(s) file:///boot/efi/EFI/centos/shim.efi
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/errors.rb:157:in `fail'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type/file/source.rb:190:in `metadata'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type/file/source.rb:126:in `copy_source_values'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type/file.rb:816:in `retrieve'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/type.rb:1133:in `retrieve_resource'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction/resource_harness.rb:307:in `from_resource'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction/resource_harness.rb:20:in `evaluate'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:267:in `apply'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:287:in `eval_resource'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:191:in `call'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:191:in `block (2 levels) in evaluate'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:521:in `block in thinmark'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/2.5.0/benchmark.rb:308:in `realtime'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:520:in `thinmark'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:191:in `block in evaluate'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/graph/relationship_graph.rb:122:in `traverse'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction.rb:178:in `evaluate'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/catalog.rb:240:in `block (2 levels) in apply'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:521:in `block in thinmark'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/2.5.0/benchmark.rb:308:in `realtime'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:520:in `thinmark'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/catalog.rb:239:in `block in apply'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/log.rb:161:in `with_destination'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/transaction/report.rb:146:in `as_logging_destination'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/catalog.rb:238:in `apply'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:185:in `block (2 levels) in apply_catalog'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:521:in `block in thinmark'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/2.5.0/benchmark.rb:308:in `realtime'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:520:in `thinmark'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:184:in `block in apply_catalog'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:233:in `block in benchmark'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/2.5.0/benchmark.rb:308:in `realtime'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:232:in `benchmark'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:183:in `apply_catalog'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:387:in `run_internal'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:227:in `block in run'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/context.rb:62:in `override'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet.rb:290:in `override'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:210:in `run'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/apply.rb:343:in `apply_catalog'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/apply.rb:260:in `block (2 levels) in main'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/context.rb:62:in `override'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet.rb:290:in `override'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/apply.rb:243:in `block in main'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/context.rb:62:in `override'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet.rb:290:in `override'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/apply.rb:207:in `main'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/apply.rb:177:in `run_command'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:382:in `block in run'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:710:in `exit_on_fail'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:382:in `run'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:143:in `run'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:77:in `execute'
    [ERROR 2020-04-20T19:43:51 verbose] /opt/puppetlabs/puppet/bin/puppet:5:in `<main>'
```

**Comment by ehelms** [comment]
I opened https://github.com/theforeman/puppet-foreman_proxy/pull/583 to tackle specific runtime issue I see for El8 support.

**Comment by wbclark** [comment]
@ehelms latest changes I've pushed here should fix the el8 spec tests. Investigating failures on centos{7,8} acceptance tests next.

**Comment by wbclark** [comment]
@ehelms @ekohl could you point to the repository that is no longer enabled by the module, so that I can re-enable it for the acceptance tests here?

**Comment by ekohl** [comment]
https://github.com/theforeman/puppet-foreman/pull/830 would fix that. Haven't had time to dive into the test failures.

**Comment by ekohl** [comment]
https://github.com/theforeman/puppet-foreman/pull/831

File: spec/classes/foreman_proxy__tftp__netboot_spec.rb [hunk 2]
```diff
@@ -50,6 +52,18 @@
             it { should contain_file("/tftproot/grub2/grubx64.efi").with_source('/boot/efi/EFI/centos/grubx64.efi') }
             it { should contain_file("/tftproot/grub2/shim.efi").with_source('/boot/efi/EFI/centos/shim.efi').with_owner('root').with_mode('0644') }
           end
+        elsif facts[:operatingsystemmajrelease].to_i == 8
```
Comment (ekohl) [inline spec/classes/foreman_proxy__tftp__netboot_spec.rb:55]: I'd test with `>=`
```suggestion
        elsif facts[:operatingsystemmajrelease].to_i >= 8
```

File: spec/classes/foreman_proxy__tftp__netboot_spec.rb [hunk 3]
```diff
@@ -50,6 +52,18 @@
...
+            it { should contain_file("/tftproot/grub2/shimx64.efi").with_source('/boot/efi/EFI/redhat/shimx64.efi').with_owner('root').with_mode('0644') }
+          when 'Fedora'
+            it { should contain_file("/tftproot/grub2/grubx64.efi").with_source('/boot/efi/EFI/fedora/grubx64.efi') }
+            it { should contain_file("/tftproot/grub2/shimx64.efi").with_source('/boot/efi/EFI/fedora/shimx64.efi').with_owner('root').with_mode('0644') }
+          when 'CentOS'
+            it { should contain_file("/tftproot/grub2/grubx64.efi").with_source('/boot/efi/EFI/centos/grubx64.efi') }
+            it { should contain_file("/tftproot/grub2/shimx64.efi").with_source('/boot/efi/EFI/centos/shimx64.efi').with_owner('root').with_mode('0644') }
+          end
         else
           it { is_expected.to contain_class('foreman_proxy::tftp::netboot').with_grub_installation_type('redhat_old') }
```
Comment (ekohl) [inline spec/classes/foreman_proxy__tftp__netboot_spec.rb:68]: Perhaps this is a good time to drop the EL6 code since this module doesn't support EL6. We also don't support running on EL 7.4 so all of that can be cleaned up.

**Comment by wbclark** [comment]
Currently it's failing for me due to:

```
An error occurred in a `before(:suite)` hook.
Failure/Error: apply_manifest_on(host, File.read(local_setup), catch_failures: true)
Beaker::Host::CommandFailure:
  Host 'centos7-64-1' exited with 1 running:
   puppet apply --verbose --detailed-exitcodes /tmp/apply_manifest.pp.GLFwG8
  Last 10 lines of output were:
  	Error: Evaluation Error: Error while evaluating a Resource Statement, Could not find declared class foreman::repo (file: /tmp/apply_manifest.pp.GLFwG8, line: 1, column: 1) on node centos7-64-1
```

I believe we are using a module in `spec/setup_acceptance_node.pp` that wasn't yet defined as a dependency.

File: spec/setup_acceptance_node.pp [hunk 4]
```diff
@@ -1,3 +1,11 @@
...
   repo => 'nightly',
 }
+
+if $facts['os']['name'] == 'CentOS' {
+  if $facts['os']['release']['major'] == 7 {
+    package { 'centos-release-scl':
+      ensure => present,
+    }
+  }
+}
```
Comment (ekohl) [inline spec/setup_acceptance_node.pp:11]: This isn't needed, `foreman::repo` does this.

**Comment by ekohl** [comment]
Did you run `spec_prep`?

**Comment by wbclark** [comment]
`spec_prep` resolved that, now I'm getting:

```
centos7-64-1 17:32:44$ puppet apply --verbose --detailed-exitcodes /tmp/apply_manifest.pp.WTdHKE
  Info: Loading facts
  Info: Loading facts
  Info: Loading facts
  Info: Loading facts
  Info: Loading facts
  Info: Loading facts
  Info: Loading facts
  Info: Loading facts
  Error: Evaluation Error: Error while evaluating a Function Call, 'upcase' parameter 'arg' expects a value of type Numeric, String, Array, or Hash, got Undef (file: /etc/puppetlabs/code/environments/production/modules/foreman_proxy/manifests/params.pp, line: 279, column: 29) on node centos7-64-1

centos7-64-1 executed in 3.10 seconds
Exited: 1
  applies with no errors (FAILED - 1)
```

I will check why that's happening.

File: spec/setup_acceptance_node.pp (hunk 4 above)
Comment (wbclark) [reply spec/setup_acceptance_node.pp]: Got it, thanks!

File: spec/classes/foreman_proxy__tftp__netboot_spec.rb (hunk 3 above)
Comment (wbclark) [reply spec/classes/foreman_proxy__tftp__netboot_spec.rb:68]: I will have to research this further. I assume it involves changes to multiple manifests and tests.

**Comment by wbclark** [comment]
So, tests passed on Travis but the centos7 test is still failing for me locally due to the `domain` fact not being defined:

```
[vagrant@centos7-64-1 ~]$ facter networking | egrep 'fqdn|hostname|domain'
  fqdn => "centos7-64-1",
  hostname => "centos7-64-1",
[vagrant@centos7-64-1 ~]$ 
```

I think we could fix that just by making `$dns_zone` default to `example.com` when `$facts['networking']['domain']` is undefined.

**Comment by wbclark** [comment]
Confirmed `spec/acceptance/basic_spec.rb` is also failing on local el8 for the same reason.

**Comment by ekohl** [comment]
IMHO it's still better to start at the lower tiers, then work your way up. https://github.com/theforeman/foreman-installer-modulesync#finding-the-order

You're now testing a module that depends on modules that themselves don't declare themselves compatible with EL8 and that can make you run into issues.

File: spec/classes/foreman_proxy__tftp__netboot_spec.rb (hunk 3 above)
Comment (ekohl) [reply spec/classes/foreman_proxy__tftp__netboot_spec.rb:68]: https://github.com/theforeman/puppet-foreman_proxy/pull/590

**Comment by ehelms** [comment]
Looks like this just needs two tests updated due to differences between EL7 and EL8

**Comment by wbclark** [comment]
> Looks like this just needs two tests updated due to differences between EL7 and EL8

Yeah, I lost the fix there when rebasing. It's my next priority when I'm off this call :)

**Comment by ehelms** [comment]
Failures look related to a new release of beaker today: https://rubygems.org/gems/beaker/versions/4.22.0

**Comment by wbclark** [comment]
pinning beaker <= 4.22.0 appears to have resolved it. Updated PR title since this is no longer WIP, will update commit messages and push once more, then this should be good to go I think.

File: spec/classes/foreman_proxy__tftp__netboot_spec.rb [hunk 5]
```diff
@@ -41,13 +41,21 @@
         case facts[:operatingsystem]
         when /^(RedHat|Scientific|OracleLinux)$/
           it { should contain_file("/tftproot/grub2/grubx64.efi").with_source('/boot/efi/EFI/redhat/grubx64.efi') }
-          it { should contain_file("/tftproot/grub2/shim.efi").with_source('/boot/efi/EFI/redhat/shim.efi').with_owner('root').with_mode('0644') }
+          if facts[:os]['release']['major'].to_i >= 8
+            it { should contain_file("/tftproot/grub2/shimx64.efi").with_source('/boot/efi/EFI/redhat/shimx64.efi').with_owner('root').with_mode('0644') }
```
Comment (mmoll) [inline spec/classes/foreman_proxy__tftp__netboot_spec.rb:45]: is this related to GH-592?

Comment (wbclark) [reply spec/classes/foreman_proxy__tftp__netboot_spec.rb:45]: Yes, good catch. The other PR is to essentially drop shim.efi and use only shimx64.efi (both are available on EL7). If that is merged, it would simplify the tests here because there would no longer be a difference between EL7 and EL8.

I think we could go ahead and merge this, since the other one is waiting on contrib with a few requested changes, and the other one would then also need to recombine the EL{7,8} tests to simply test for shimx64.efi in all cases. What do you think @mmoll ?

Comment (mmoll) [reply spec/classes/foreman_proxy__tftp__netboot_spec.rb:45]: I'd prefer to have other PR in first, tbh...

File: Gemfile [hunk 6]
```diff
@@ -5,6 +5,7 @@ source 'https://rubygems.org'
 
 gem 'puppet', ENV.key?('PUPPET_VERSION') ? "~> #{ENV['PUPPET_VERSION']}" : '>= 5.5'
 
+gem 'beaker', '< 4.22.0'
```
Comment (ekohl) [inline Gemfile:8]: This is a temporary workaround that's no longer needed.

**Comment by ekohl** [comment]
Please rebase

**Comment by ehelms** [comment]
Thanks @wbclark !
